src/nvim/runtime.c

```diff
@@ -1841,6 +1845,183 @@ static void cmd_source_buffer(const exarg_T *const eap)
   ga_clear(&ga);
 }
 
+/// Read a file, prompting the user if the file should be trusted or not.
+///
+/// Maintain a "trust database" at stdpath('state')/trust containing a sequence
+/// of [path, hash] pairs. A path can be marked as blacklisted by using a '!'
+/// character for the hash.
+///
+/// @param path Path to file
+///
+/// @return A pointer to file contents if the file should be trusted, or NULL
+///         otherwise. Caller owns returned memory.
+const char *read_secure(const char *path)
+  FUNC_ATTR_NONNULL_ALL
+{
+  char *contents = NULL;
+
+  // Exit early if file can't be read
+  FILE *fp = fopen(path, "r");
+  if (!fp) {
+    return NULL;
+  }
+
+  PMap(cstr_t) file_hashes = MAP_INIT;
+  char *fullpath = os_realpath(path, NULL);
+  char *trustdb = stdpaths_user_state_subpath("trust", 0, false);
+  FILE *trust_fp = os_fopen(trustdb, "r");
+  if (trust_fp) {
+    char *line = NULL;
+    size_t len = 0;
+    while ((line = fgetln(trust_fp, &len)) != NULL) {
+      char *p = strsep(&line, " \n");
+      char *h = strsep(&line, " \n");
+      if (p == NULL || h == NULL) {
+        // Malformed line
+        continue;
+      }
+      (void)pmap_put(cstr_t)(&file_hashes, p, xstrdup(h));
+    }
+
+    fclose(trust_fp);
+  }
+
+  const char *curhash = pmap_get(cstr_t)(&file_hashes, fullpath);
+  if (curhash != NULL && strcmp(curhash, "!") == 0) {
```

Comment:
The filename may actually start with a `!` character which will need to be escaped when it's added to the file.

Reply:
I believe this is the hash column, not path, so there is no conflict here.

Reply:
That is correct.